crypto/fipsmodule/rand/ctrdrbg.c

```diff
@@ -26,65 +26,115 @@
+      init_mask = kInitMask128;
+      break;
+    case CTR_DRBG_AES_256_KEY_LEN:
+      drbg->aes_key_len = CTR_DRBG_AES_256_KEY_LEN;
+      drbg->entropy_len = CTR_DRBG_AES_256_ENTROPY_LEN;
+      init_mask = kInitMask256;
+      break;
+    default:
+      return 0;
   }
 
-  drbg->ctr = aes_ctr_set_key(&drbg->ks, NULL, &drbg->block, seed_material, 32);
-  OPENSSL_memcpy(drbg->counter.bytes, seed_material + 32, 16);
-  drbg->reseed_counter = 1;
+  // Section 10.2.1.3.1
+  if (personalization_len > drbg->entropy_len) {
+    return 0;
+  }
+
+  uint8_t seed_material[CTR_DRBG_MAX_ENTROPY_LEN];
+  OPENSSL_memcpy(seed_material, entropy, drbg->entropy_len);
+
+  for (size_t i = 0; i < personalization_len; i++) {
+    seed_material[i] ^= personalization[i];
+  }
+
+  // Section 10.2.1.2
+  for (size_t i = 0; i < drbg->entropy_len; i++) {
+    seed_material[i] ^= init_mask[i];
+  }
+
+  drbg->ctr = aes_ctr_set_key(&drbg->ks, NULL, &drbg->block, seed_material,
+    drbg->aes_key_len);
 
+  OPENSSL_STATIC_ASSERT(sizeof(drbg->counter.bytes) >= CTR_DRBG_STATE_COUNTER_LEN_IN_BYTES,
+    CTR_DRBG_state_counter_insufficient_size)
+  OPENSSL_memcpy(drbg->counter.bytes, seed_material + drbg->aes_key_len,
+    CTR_DRBG_STATE_COUNTER_LEN_IN_BYTES);
+
+  drbg->reseed_counter = 1;
   return 1;
 }
 
-OPENSSL_STATIC_ASSERT(CTR_DRBG_ENTROPY_LEN % AES_BLOCK_SIZE == 0,
-                      not_a_multiple_of_AES_block_size)
+// Simplifies 10.2.1.2 and 10.2.1.5.1.
+OPENSSL_STATIC_ASSERT(CTR_DRBG_AES_128_ENTROPY_LEN % AES_BLOCK_SIZE == 0,
+  not_a_multiple_of_AES_block_size)
+OPENSSL_STATIC_ASSERT(CTR_DRBG_AES_256_ENTROPY_LEN % AES_BLOCK_SIZE == 0,
+  not_a_multiple_of_AES_block_size)
+OPENSSL_STATIC_ASSERT(CTR_DRBG_STATE_COUNTER_LEN_IN_BYTES == AES_BLOCK_SIZE,
+  state_counter_length_is_not_equal_to_the_block_length)
 
 // ctr_inc adds |n| to the last four bytes of |drbg->counter|, treated as a
 // big-endian number.
 static void ctr32_add(CTR_DRBG_STATE *drbg, uint32_t n) {
+  OPENSSL_STATIC_ASSERT(sizeof(drbg->counter.words) == (4 * sizeof(uint32_t)),
+    ctr_drbg_counter_is_not_of_expected_size)
+
   drbg->counter.words[3] =
       CRYPTO_bswap4(CRYPTO_bswap4(drbg->counter.words[3]) + n);
 }
 
+// |drbg| is not |NULL| when this function is called.
+// |data_len| is <= |CTR_DRBG_MAX_ENTROPY_LEN|.
 static int ctr_drbg_update(CTR_DRBG_STATE *drbg, const uint8_t *data,
                            size_t data_len) {
-  // Per section 10.2.1.2, |data_len| must be |CTR_DRBG_ENTROPY_LEN|. Here, we
+  // Per section 10.2.1.2, |data_len| must be |drbg->entropy_len|. Here, we
   // allow shorter inputs and right-pad them with zeros. This is equivalent to
   // the specified algorithm but saves a copy in |CTR_DRBG_generate|.
-  if (data_len > CTR_DRBG_ENTROPY_LEN) {
+  if (data_len > drbg->entropy_len) {
     return 0;
   }
 
-  uint8_t temp[CTR_DRBG_ENTROPY_LEN];
-  for (size_t i = 0; i < CTR_DRBG_ENTROPY_LEN; i += AES_BLOCK_SIZE) {
+  uint8_t temp[CTR_DRBG_MAX_ENTROPY_LEN];
```

Comment:
Does temp need to be initialised to 0 or it doesn't matter? I see it wasn't in the original code.

Reply:
To be precise. `drbg->entropy_len` bytes of `tmp` is overwritten, and these are the only bytes we care about.